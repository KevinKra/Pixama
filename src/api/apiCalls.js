import API_KEY from "./apikey";

export const fetchPopularMovies = async () => {
  let response = await fetch(
    `https://api.themoviedb.org/3/discover/movie?certification_country=US&api_key=${API_KEY}`
  );
  const movies = await response.json();
  return movies.results;
};

//This method is not hooked up anywhere at this point
//data must be an object with keys of name, email and password
export const fetchNewUser = (url, data)  => {
  return fetch(url, 
    {
      method: 'POST',
      body: JSON.stringify( data ),
      headers: {
        'Content-Type': 'application/json'
    }
  })
  .then(response=> {
    if(!response.ok) {
      throw new Error ("Email already exists", response.message)
    } else {
      return response.json()
    }
  })
}

//this method is being imported and called in LoginCard component
export const fetchUser = (url, data)  => {

  return fetch(url, 
    {
      method: 'POST',
      body: JSON.stringify( data ),
      headers: {
        'Content-Type': 'application/json'
    }
  })
  .then(response=> {
    if(!response.ok) {
      throw new Error ("Not a valid login",response.message)
    } else {
      return response.json()
    }
  })
}

//Method to add a movie to a user's favorites, this method not hooked up to a component yet
// url in component method needs to be 'http://localhost:3000/api/users/favorites/new'
//data needs to be object like this: {movie_id: 550, user_id: 1, title: "Fight Club", poster_path: "/adw6Lq9FiC9zjYEpOqfq03ituwp.jpg" , release_date: "1999-10-15", vote_average:8.4, overview: "A ticking-time-bomb insomniac and a slippery soap salesman channel primal male aggression into a shocking new form of therapy. Their concept catches on, with underground \"fight clubs\" forming in every town, until an eccentric gets in the way and ignites an out-of-control spiral toward oblivion."}
//response will be the new favorite ID
export const postFavorite = (url, data)  => {
console.log(data)
return fetch(url, 
    {
      method: 'POST',
      body: JSON.stringify( data ),
      headers: {
        'Content-Type': 'application/json'
    }
  })
  .then(response=> {
    if(!response.ok) {
      throw new Error ("Add error",response.message)
    } else {
      return response.json()
    }
  })
};


export const fetchFavorites = (url)  => {

return fetch(url)
  .then(response=> {
    if(!response.ok) {
      throw Error ("Couldn't get favorites",response.message)
    } else {
      return response.json()
    }
  })
};