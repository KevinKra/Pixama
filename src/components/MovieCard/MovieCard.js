import React, { Component, Fragment } from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faBookmark } from "@fortawesome/free-solid-svg-icons";
import { faPlay } from "@fortawesome/free-solid-svg-icons";
import {
  postFavorite,
  fetchFavorites,
  deleteFavorite
} from "../../api/apiCalls";
import { connect } from "react-redux";
import { updatePopularFavorites, updateRomanceFavorites } from "../../actions";
import "./MovieCard.scss";

export class MovieCard extends Component {
  state = {
    displayBackdrop: false,
    userActive: false,
    bookmarked: this.props.isFavorite 
  };

  displayBackdrop = () => {
    setTimeout(this.determineActive, 550);
    this.setState({ userActive: true });
  };

  displayPoster = () => {
    this.setState({ displayBackdrop: false, userActive: false });
  };

  determineActive = () => {
    this.state.userActive
      ? this.setState({ displayBackdrop: true })
      : this.setState({ displayBackdrop: false });
  };

  bookmarkCard = async () => {
    // const toggle = this.state.bookmarked;
    // this.setState({ bookmarked: !toggle });
    if (!this.props.isFavorite) {
      const {
        id,
        title,
        poster,
        releaseDate,
        popularity,
        overview,
        currentUser
      } = this.props;
      const data = {
        movie_id: id,
        user_id: currentUser.id,
        title,
        poster_path: poster,
        release_date: releaseDate,
        vote_average: popularity,
        overview
      };
      await postFavorite("http://localhost:3000/api/users/favorites/new", data);
      const favorites = await fetchFavorites(
        `http://localhost:3000/api/users/${this.props.currentUser.id}/favorites`
      );
      const favoriteIDs = favorites.data.map(favorite => favorite.movie_id)
      const popularFavorites = this.props.popularMovies.map(movie => {
        return favoriteIDs.includes(movie.id) ? { ...movie, isFavorite: true } : movie;
      });

      this.props.updatePopularFavorites(popularFavorites);

      const romanceFavorites = this.props.romanceMovies.map(movie => {
        return favoriteIDs.includes(movie.id) ? { ...movie, isFavorite: true } : movie;
      });
      this.props.updateRomanceFavorites(romanceFavorites); 
    } else {
      await deleteFavorite(
        `http://localhost:3000/api/users/${
          this.props.currentUser.id
        }/favorites/${this.id}`
      );
      
      const favorites = await fetchFavorites(
        `http://localhost:3000/api/users/${
          this.props.currentUser.id
        }/favorites`
      );
      const favoriteIDs = favorites.data.map(
        favorite => favorite.movie_id
      );
      const popularFavorites = this.props.popularMovies.map(movie => {
        return favoriteIDs.includes(movie.id) ? { ...movie, isFavorite: true } : movie;
      });

      this.props.updatePopularFavorites(popularFavorites);

      const romanceFavorites = this.props.romanceMovies.map(movie => {
        return favoriteIDs.includes(movie.id) ? { ...movie, isFavorite: true } : movie;
      });
      this.props.updateRomanceFavorites(romanceFavorites); 
    }
  };

  render() {
    const bookmark = (
      <FontAwesomeIcon
        icon={faBookmark}
        size="lg"
        onClick={this.bookmarkCard}
        className={`bookmark-icon ${this.props.isFavorite && "bookmarked"}`}
      />
    );

    const play = (
      <FontAwesomeIcon
        icon={faPlay}
        size="lg"
        onClick={() => console.log("play btn clicked")}
        className="play-icon"
      />
    );
    const backdrop = (
      <Fragment>
        <div className="overlay-content">
          <h3 className="movie-title">{this.props.title}</h3>
          <p className="movie-overview">{this.props.overview}</p>
          {bookmark}
          {play}
          <p className="movie-language">{`Language: ${this.props.language.toUpperCase()}`}</p>
          <p className="movie-popularity">{`Popularity: ${
            this.props.popularity
          }`}</p>
        </div>
        <div className="screen-solid" />
        <div
          className="backdrop-image"
          style={{
            backgroundImage: `url(https://image.tmdb.org/t/p/w780/${
              this.props.backdrop
            })`
          }}
        />
      </Fragment>
    );

    const poster = (
      <img
        className="poster-image"
        src={`https://image.tmdb.org/t/p/w185/${this.props.poster}`}
        alt={this.props.title}
      />
    );

    return (
      <article
        className="MovieCard"
        style={
          !this.state.displayBackdrop
            ? { maxWidth: "185px" }
            : {
                maxWidth: "320px",
                transform: "scale(1.03)",
                zIndex: "5"
              }
        }
        onMouseEnter={this.displayBackdrop}
        onMouseLeave={this.displayPoster}
      >
        {!this.state.displayBackdrop ? poster : backdrop}
      </article>
    );
  }
}

export const mapStateToProps = state => ({
  currentUser: state.currentUser,
  popularMovies: state.popularMovies,
  romanceMovies: state.romanceMovies
});

export const mapDispatchToProps = dispatch => ({
  updatePopularFavorites: popularFavorites => dispatch(updatePopularFavorites(popularFavorites )),
  updateRomanceFavorites: romanceFavorites => dispatch(updateRomanceFavorites(romanceFavorites))
});

export default connect(
  mapStateToProps, mapDispatchToProps
)(MovieCard);
