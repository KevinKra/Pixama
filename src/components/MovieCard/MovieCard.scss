@import "../../_styles//variables.scss";

.MovieCard {
  position: relative;
  border: 1px solid rgb(15, 15, 15);
  z-index: 3;
  height: 278px;
  margin: 0 0.3rem;
  transition: 0.4s min-width ease-in, 0.35s transform ease-in;
  cursor: pointer;
  .poster-image {
    position: absolute;
    top: 0;
    left: 0;
  }
  .poster-name {
    position: absolute;
    animation: fade-in-hold 1.5s forwards;
    z-index: 3;
    left: 15px;
    color: $text-light;
  }
  .backdrop-image {
    background-size: cover;
    background-position: center;
    height: 100%;
  }
}

@keyframes fade-in-hold {
  0% {
    opacity: 0;
  }
  30% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}
