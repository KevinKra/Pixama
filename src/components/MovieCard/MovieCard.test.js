import { MovieCard } from "./MovieCard";
import React from "react";
import { shallow } from "enzyme";
import * as apiCalls from "../../api/apiCalls";

describe("MovieCard", () => {
  let wrapper;
  let instance;
  let props = {
    updatePopularFavorites: jest.fn(),
    updateRomanceFavorites: jest.fn(),
    updateMoviePage: jest.fn(),
    updateFavorites: jest.fn(),
    popularMovies: [{ id: 1 }, { id: 2 }],
    romanceMovies: [{ id: 3 }, { id: 4 }],
    favorites: [{ id: 5 }, { id: 6 }],
    isFavorite: false,
    id: 3,
    history: [],
    currentUser: { id: 5, loggedIn: true },
    title: "movie title",
    poster: "path",
    releaseDate: "11/25/1989",
    popularity: 7.2,
    overview: "its good",
    backdrop: "string",
    language: "english"
  };

  beforeEach(() => {
    wrapper = shallow(<MovieCard {...props} />);
    instance = wrapper.instance();
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it.skip("should call fetchMovies twice on mount", async () => {
    // instance.componentDidMount();
    const popMovies = await apiCalls.fetchMovies("");
    const romMovies = await apiCalls.fetchMovies("");
    expect(apiCalls.fetchMovies).toHavevBeenCalledTimes(2);
  });
});
