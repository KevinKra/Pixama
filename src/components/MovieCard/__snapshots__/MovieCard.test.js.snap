// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard should match the snapshot 1`] = `
<article
  className="MovieCard"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "minWidth": "185px",
    }
  }
>
  <img
    alt="movie title"
    className="poster-image notBookmarked"
    onClick={[Function]}
    src="https://image.tmdb.org/t/p/w185/path"
  />
</article>
`;
